integration/users/create_user_test.go

```diff
@@ -0,0 +1,185 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package users
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+
+	"github.com/FerretDB/FerretDB/integration"
+	"github.com/FerretDB/FerretDB/integration/setup"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+	"github.com/FerretDB/FerretDB/internal/util/testutil"
+)
+
+func TestCreateUser(t *testing.T) {
+	t.Parallel()
+
+	ctx, collection := setup.Setup(t)
+	db := collection.Database()
+	client := db.Client()
+	users := client.Database("admin").Collection("system.users")
+
+	// TODO https://github.com/FerretDB/FerretDB/issues/1492
+	if setup.IsMongoDB(t) {
+		assert.NoError(t, collection.Database().RunCommand(ctx, bson.D{
+			{"dropAllUsersFromDatabase", 1},
+		}).Err())
+	} else {
+		// Erase any previously saved user in the database.
+		_, err := users.DeleteMany(ctx, bson.D{{"db", db.Name()}})
+		assert.NoError(t, err)
+	}
+
+	testCases := map[string]struct { //nolint:vet // for readability
+		payload    bson.D
+		err        *mongo.CommandError
+		altMessage string
+		expected   bson.D
+		skip       string
+	}{
+		"AlreadyExists": {
+			payload: bson.D{
+				{"createUser", "should_already_exist"},
+				{"roles", bson.A{}},
+				{"pwd", "password"},
+			},
+			err: &mongo.CommandError{
+				Code:    51003,
+				Name:    "Location51003",
+				Message: "User \"should_already_exist@TestCreateUser\" already exists",
+			},
+		},
+		"MissingPwdOrExternal": {
+			payload: bson.D{
+				{"createUser", "mising_pwd_or_external"},
+				{"roles", bson.A{}},
+			},
+			err: &mongo.CommandError{
+				Code:    2,
+				Name:    "BadValue",
+				Message: "Must provide a 'pwd' field for all user documents, except those with '$external' as the user's source db",
+			},
+		},
+		"Success": {
+			payload: bson.D{
+				{"createUser", "success_user"},
+				{"roles", bson.A{}},
+				{"pwd", "password"},
+			},
+			expected: bson.D{
+				{
+					"ok", float64(1),
+				},
+			},
+		},
+		"WithComment": {
+			payload: bson.D{
+				{"createUser", "with_comment_user"},
+				{"roles", bson.A{}},
+				{"pwd", "password"},
+				{"comment", "test string comment"},
+			},
+			expected: bson.D{
+				{
+					"ok", float64(1),
+				},
+			},
+		},
+		"MissingRoles": {
+			payload: bson.D{
+				{"createUser", "missing_roles"},
+				{"pwd", "password"},
+			},
+			err: &mongo.CommandError{
+				Code:    40414,
+				Name:    "Location40414",
+				Message: "BSON field 'createUser.roles' is missing but a required field",
+			},
+		},
+	}
```

Comment:
Are there any other validation-related cases? E.g. user name contains invalid symbols or smth like this. Or is it out of scope for now?

Reply:
I haven't found what rules are used for validation and testing it seems to be pretty liberal, but one thing I missed was a test case for checking emptiness. Doing it now.